### PR TITLE
[vividus] Add support for vividus expressions in templates

### PR DESCRIPTION
**Pull request by uarlouski**
Relates #1545

### CODE REVIEW

**Comment by ikalinin1** [comment]
@uarlouski  please add integration test

**Comment by uarlouski** [comment]
> @uarlouski please add integration test

it's draft :)

File: vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+        @SuppressWarnings({"rawtypes", "unchecked"})
+        @Override
+        public Object exec(List arguments) throws TemplateModelException
+        {
+            String expressionName = arguments.get(0).toString();
+            String expression = (String) arguments.subList(1, arguments.size())
+                                                  .stream()
+                                                  .map(Object::toString)
+                                                  .collect(Collectors.joining(",", expressionName + "(", ")"));
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java:129]: what about nested expressions?

File: vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+            return processors.stream().map(p -> p.execute(expression))
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java:131]: I think it makes sense to use `ExpressionAdaptor` to avoid copy-pasted logic

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java (hunk 1 above)
Comment (uarlouski) [reply vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java:129]: how do you imagine it will look like?

Comment (uarlouski) [reply vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java:129]: added example

Comment (valfirst) [reply vividus/src/main/java/org/vividus/bdd/steps/FreemarkerSteps.java:129]: please update the docs as well

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ExpressionAdaptor.java [hunk 3]
```diff
@@ -41,12 +42,61 @@
 
     private List<IExpressionProcessor<?>> processors;
 
-    public Object process(String value)
+    /**
+     * Processes the expression including processing nested ones.
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ExpressionAdaptor.java:46]: `Processes the expression including nested ones.`

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ExpressionAdaptor.java [hunk 4]
```diff
@@ -41,12 +42,61 @@
...
+     * @return the result of processed expression
+     */
+    public Object processRawExpression(String expression)
+    {
+        return wrap(expression, () -> processExpression(expression,
+                List.of(RELUCTANT_EXPRESSION_PATTERN, GREEDY_EXPRESSION_PATTERN).iterator()));
+    }
+
+    /**
+     * Processes the expression without processing nested ones.
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ExpressionAdaptor.java:73]: `Processes the expression excluding nested ones.`

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ExpressionAdaptor.java [hunk 5]
```diff
@@ -41,12 +42,61 @@
...
+     *
+     * @param expression the expression to process
+     * @return the result of processed expression
+     */
+    public Object processExpression(String expression)
+    {
+        return wrap(expression, () -> apply(expression));
+    }
+
+    private Object wrap(String value, Supplier<Object> expressionResolver)
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/steps/ExpressionAdaptor.java:95]: `wrap` -> `processExpression`

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Pull request merged**